staging/src/k8s.io/apiserver/pkg/endpoints/handlers/apply/encode.go

```diff
@@ -0,0 +1,193 @@
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// EncodeManagedFields converts ManagedFields from the the format used by
+// https://github.com/kubernetes-sigs/structured-merge-diff to the the wire format (api format)
+func EncodeManagedFields(managedFields fieldpath.ManagedFields) (encodedManagedFields map[string]metav1.VersionedFieldSet, err error) {
+	encodedManagedFields = make(map[string]metav1.VersionedFieldSet, len(managedFields))
+	for manager, versionedSet := range managedFields {
+		v, err := encodeVersionedSet(versionedSet)
+		if err != nil {
+			return nil, fmt.Errorf("error encoding versioned set for %v: %v", manager, err)
+		}
+		encodedManagedFields[manager] = *v
+	}
+	return encodedManagedFields, nil
+}
+
+func encodeVersionedSet(versionedSet *fieldpath.VersionedSet) (encodedVersionedSet *metav1.VersionedFieldSet, err error) {
+	encodedVersionedSet = &metav1.VersionedFieldSet{}
+	encodedVersionedSet.APIVersion = string(versionedSet.APIVersion)
+	f, err := encodeSet(versionedSet.Set)
+	if err != nil {
+		return nil, fmt.Errorf("error encoding set: %v", err)
+	}
+	encodedVersionedSet.Fields = *f
+	return encodedVersionedSet, nil
+}
+
+func encodeSet(set *fieldpath.Set) (encodedSet *metav1.FieldSet, err error) {
+	encodedSet = &metav1.FieldSet{}
+	encodedSet.Members, err = encodePathElementSet(&set.Members)
+	if err != nil {
+		return nil, fmt.Errorf("error encoding path element set: %v", err)
+	}
+	encodedSet.Children, err = encodeSetNodeMap(&set.Children)
+	if err != nil {
+		return nil, fmt.Errorf("error encoding children: %v", err)
+	}
+	return encodedSet, nil
+}
+
+func encodePathElementSet(pathElementSet *fieldpath.PathElementSet) (encodedPathElementSet []metav1.FieldPathElement, err error) {
+	// Get a list of the elements in the pathElementSet and sort it
+	pathElementList := make(sortablePathElementList, pathElementSet.Size())
+	i := 0
+	pathElementSet.Iterate(func(pathElement fieldpath.PathElement) {
+		pathElementList[i] = &pathElement
+		i++
+	})
+	sort.Sort(pathElementList)
+
+	encodedPathElementSet = make([]metav1.FieldPathElement, len(pathElementList))
+	for i, pathElement := range pathElementList {
+		encodedPathElement, err := encodePathElement(pathElement)
+		if err != nil {
+			return nil, fmt.Errorf("error encoding path element: %v", err)
+		}
+		encodedPathElementSet[i] = encodedPathElement
+	}
+	return encodedPathElementSet, nil
+}
+
+func encodeSetNodeMap(setNodeMap *fieldpath.SetNodeMap) (encodedSetNodeMap []metav1.FieldSetNode, err error) {
+	// This is the only way to get the number of keys in the setNodeMap,
+	// the Size() function will include the size of all the sub paths
+	count := 0
+	setNodeMap.Iterate(func(_ fieldpath.PathElement) {
+		count++
+	})
+
+	// Get a list of the keys in the setNodeMap and sort it
+	pathElementList := make(sortablePathElementList, count)
+	i := 0
+	setNodeMap.Iterate(func(pathElement fieldpath.PathElement) {
+		pathElementList[i] = &pathElement
```

Comment:
I recommend just using append() instead of precomputing the count.

(If we're worried about the number of allocations then copying pathElement instead of taking its address is probably more important, anyway. But we shouldn't worry about this right now.)